### PR TITLE
Improve README

### LINKED ISSUE #8: needs a better readme file
@slawrence87544 can you make a quick PR to the readme file that describes how to use this repo (entry point. rdbms config, etc)

### PR DESCRIPTION
**Pull request by jrknezha**
Improvements for the README to include more specific documentations, particularly request dictionary examples and code structure descriptions. 

A few additional text changes are made in other files which were identified as part of the new README documentations for clarity and consistency.

### CODE REVIEW

File: src/db_request_registry.py [hunk 1]
```diff
@@ -47,7 +47,7 @@
         DbActionResponse
     ),
     'expt_metrics': RequestHandler(
-        'Add or get or update experiment metrics data',
+        'Add or get experiment metrics data',
         ExptMetricRequest,
```
Comment (amschne) [inline src/db_request_registry.py:51]: Will you please comment on the reason for this change? I notice the same descriptions throughout this dictionary ("'Add or get or update...") that remain unchanged.

File: README.md [hunk 2]
```diff
@@ -73,7 +73,26 @@ Each of the APIs is structured in a similar way and are meant to be
...
 with a yaml file or a python dictionary as the only accepted arguments.
 
-#### How to Register an Experiment
+## Input values 
+Every action made through the API requires a name and method. Valid methods are GET or PUT. The name must be a registered value in the db_request_regsitry. 
+Currently valid registry options are:
+```sh
+'region' : 'Add or get regions'
+'experiment' : 'Add, get, or update experiment registration data'
+'expt_metrics' : 'Add or update experiment metrics data'
```
Comment (amschne) [inline README.md]: noting here that the corresponding string for expt_metrics in the source file is "Add or [get]..."

File: README.md [hunk 3]
```diff
@@ -73,7 +73,26 @@ Each of the APIs is structured in a similar way and are meant to be
...
+'metric_types' : 'Add, get, or update metric types'
+'harvest_innov_stats' : 'Gather and store innovation statistics from diagnostics files'
+'plot_innov_stats' : 'Plot innovation statistics'
+'file_types' : 'Add, get, or update file types'
+'storage_locations' : 'Add, get, or update storage locations'
+'expt_file_counts' : 'Add or get experiment file counts'
+'harvest_metrics' : 'Harvest and store metrics'
+```
+
+Example request dictionaries for each registry option are provided in the index.
```
Comment (amschne) [inline README.md]: Very helpful, thank you! I think maybe "index" should instead be "Appendix" or even something simple like "below." I guess I'm just looking for something a little more clear.

File: README.md [hunk 4]
```diff
@@ -439,7 +462,7 @@ file_types
 
     id = Column(Integer, primary_key=True, autoincrement=True)
     name = Column(String(128), nullable=False)
-    file_template= Column(String(64), nullable=False)
+    file_template = Column(String(64), nullable=False)
```
Comment (amschne) [inline README.md:465]: wow, nice catch

File: README.md [hunk 5]
```diff
@@ -459,10 +482,500 @@ expt_stored_file_counts
...
     experiment = relationship('Experiment', back_populates='file_counts')
     file_type = relationship('FileType', back_populates='file_counts')
     storage_location = relationship('StorageLocation', back_populates='file_counts')
-```
\ No newline at end of file
+```
+
+## Request Dictionaries / YAML Formats
+
+For each request, a dictionary of values or yaml file containing the same information are necessary with relevant data for that call. GET calls use filters and ordering data. PUT calls must contain the data to be put into the database. This is a list of example calls only. Each call should be customized with the appropriate input data. While the examples are in a dictionary format, all the values could be similarly input as a yaml file with the same hierachy provided by the examples.
```
Comment (amschne) [inline README.md]: This is a great start and will be so helpful for new (and current) users. I think the passage could be revised for better clarity. How about:

```suggestion
All requests require either a dictionary or a YAML file to configure the call to the database. GET calls, e.g., are requests to download subsets of data using filters that are specified with a nested dictionary or within a YAML file configuration hierarchy. PUT calls must similarly (via defining a key:value structured hierarchy) specify which data to upload to the database. Other calls must also be configured with a dictionary or YAML file. The following example configuration dictionaries (which could be similarly defined in a YAML file with the same hierarchy) are provided as templates for basic use cases.
```

File: README.md [hunk 6]
```diff
@@ -459,10 +482,500 @@ expt_stored_file_counts
...
+
+### Experiment Dictionaries
+Example format of request dictionaries for 'experiment' calls.
+
+GET: 
+
+```sh
+request_dict = {
+        'name': 'experiment',
```
Comment (amschne) [inline README.md:505]: nice use of indentation - very clear for demonstrating the nesting levels

File: README.md [hunk 7]
```diff
@@ -459,10 +482,500 @@ expt_stored_file_counts
...
+                'forecast_hour' : forecast_hour,
+                'ensemble_member' : ensemble_member
+            },
+            'datestr_format': '%Y-%m-%d %H:%M:%S',
+        }
+    }
+```
+Values which can be null or not provided: elevation_unit, forecast_hour, ensemble_member
+
+Note: for a successful PUT call, the experiment, region, and metric type referenced in the body must already be registered.
```
Comment (amschne) [inline README.md]: I know this is specified above, but I suggest maybe adding at the end "using the db_XYZ.py" script, as demonstrated above."

File: README.md [hunk 8]
```diff
@@ -459,10 +482,500 @@ expt_stored_file_counts
...
+            'key': 'reanalysis',
+            'count': 1230,
+            'folder_path': 'noaa-example-score-db-bucket/reanalysis/2023/02/23/2023022306',
+            'cycle': '2023-02-03 06:00:00'
+        }
+    }
+```
+Values which can be null or not provided: folder_path, cycle, time_valid, forecast_hour, file_size_bytes
+
+Note: the associated experiment, file type, and storage locations referenced in the body values must already be registered for a successful file count PUT call.
```
Comment (amschne) [inline README.md]: same as above - maybe add the very short how to or which script is used to register

File: README.md [hunk 9]
```diff
@@ -459,10 +482,500 @@ expt_stored_file_counts
...
+            }
+        ],
+        'work_dir': '/absolute/path/to/desired/figure/location',
+        'fig_base_fn': 'C96L64_GSI_3DVAR_VS_GSI_SOCA_3DVAR'
+    }
+```
+
+## Code Structure
+
+The code base is developed into separate files to handling the interactions for separate tables as well as extra functionalities, i.e. harvester interactions. All of the code functionality resides with in the **src** folder.
```
Comment (amschne) [inline README.md]: it was a little difficult for me to intpret this sentence - please confirm if my suggestion below makes sense

```suggestion
The code base is structured using separate files to handle interactions for separate tables as well as extra functionalities, i.e., harvester interactions. All of the code functionality resides with in the **src** folder. 
```

File: README.md [hunk 10]
```diff
@@ -459,10 +482,500 @@ expt_stored_file_counts
...
+
+- experiments: *experiments.py*
+- regions: *regions.py*
+- metric_types: *metric_types.py*
+- expt_metrics: *expt_metrics.py*
+- storage_locations: *storage_locations.py*
+- file_types: *file_types.py*
+- expt_stored_file_counts: *expt_file_counts.py*
+
+The general structure of all of these files is to define a code structure based on the columns of the database, handle the processing of input data into the appropriate values, handling any input filters or order_by statements for GET calls, and finally to handle the GET and PUT calls using SQLAlchemy to input or retrieve data from the database.
```
Comment (amschne) [inline README.md]:
```suggestion
The general purpose of all of these files is to define a code structure based on the columns of the database, handle the processing of input data into the appropriate values, handle any input filters or order_by statements for GET calls, and finally to handle the GET and PUT calls using SQLAlchemy to input or retrieve data from the database. 
```

File: README.md [hunk 11]
```diff
@@ -459,10 +482,500 @@ expt_stored_file_counts
...
+- regions: *regions.py*
+- metric_types: *metric_types.py*
+- expt_metrics: *expt_metrics.py*
+- storage_locations: *storage_locations.py*
+- file_types: *file_types.py*
+- expt_stored_file_counts: *expt_file_counts.py*
+
+The general structure of all of these files is to define a code structure based on the columns of the database, handle the processing of input data into the appropriate values, handling any input filters or order_by statements for GET calls, and finally to handle the GET and PUT calls using SQLAlchemy to input or retrieve data from the database. 
+
+The *expt_metrics.py* and *expt_file_counts.py* files have interactions with the other codes files for the relevant tables related to the id relationships between those tables, for example both call the *experiments.py* code to get an experiment id. See the Table Schemas for the full set of table interactions.
```
Comment (amschne) [inline README.md]:
```suggestion
The *expt_metrics.py* and *expt_file_counts.py* files have interactions with the other files for the relevant tables related to the id relationships between those tables, for example, both call the *experiments.py* code to get an experiment id. See the Table Schemas for the full set of table interactions. 
```

**Review by amschne — COMMENTED** [review]
This PR adds documentation crucial for users to understand how to interact with the SQL database. In addition to important documentation, sample nested dictionary, which are required to configure a call to the database, are provided. Providing these sample dictionaries in the README represents a major improvement to the usability of score-db for new (and existing) users, which until now was difficult. I commend Jessica for this work as it will advance our efforts to train new users and develop new applications of this repository.

My only concern is that some of the new documentation can be a little hard to follow. I think it's crucial to have this documentation, but the README is now a little long. Do you think it is time to start a docs/ folder where some of this important information can live?

While reviewing, I tried to provide some meaningful suggestions to improve clarity. But if my suggestions change the meaning of anything, please rebuff as necessary.

File: src/db_request_registry.py (hunk 1 above)
Comment (jrknezha) [reply src/db_request_registry.py:51]: This update is to reflect the actual code process. The code doesn't allow updates, so I updated the documentation to reflect that. We only do additions for put calls and take the most recent version in gets to not overwrite data for metrics and file counts. The other calls in the dictionary do allow updates which is why they are unchanged.

File: README.md (hunk 2 above)
Comment (jrknezha) [reply README.md]: good catch, updated it

File: README.md (hunk 3 above)
Comment (jrknezha) [reply README.md]: updated to Appendix to match the title of the section

File: README.md (hunk 7 above)
Comment (jrknezha) [reply README.md]: added reference to the script and example above.

File: README.md (hunk 8 above)
Comment (jrknezha) [reply README.md]: added reference to the script and the example above

**Review by amschne — APPROVED** [review]
